sql/information_schema/columns_table.go

```diff
@@ -125,54 +125,106 @@ func columnsRowIter(ctx *sql.Context, cat sql.Catalog, columnNameToDefault map[s
 	for _, db := range cat.AllDatabases(ctx) {
 		// Get all Tables
 		err := sql.DBTableIter(ctx, db, func(t sql.Table) (cont bool, err error) {
+			var columnKeyMap = make(map[string]string)
+			// Get UNIQUEs, PRIMARY KEYs
+			hasPK := false
+			if indexTable, ok := t.(sql.IndexedTable); ok {
+				indexes, iErr := indexTable.GetIndexes(ctx)
+				if iErr != nil {
+					return false, iErr
+				}
+
+				for _, index := range indexes {
+					idx := ""
+					if index.ID() == "PRIMARY" {
+						idx = "PRI"
+						hasPK = true
+					} else if index.IsUnique() {
+						idx = "UNI"
+					} else {
+						idx = "MUL"
+					}
+
+					colNames := getColumnNamesFromIndex(index, t)
+					// A UNIQUE index may display as MUL if several columns form a composite UNIQUE index
+					if idx == "UNI" && len(colNames) > 1 {
+						idx = "MUL"
+						columnKeyMap[colNames[0]] = idx
+					} else {
+						for _, colName := range colNames {
+							columnKeyMap[colName] = idx
+						}
+					}
+				}
+			}
+
 			for i, c := range t.Schema() {
 				var (
 					nullable   string
+					dataType   string
 					charName   interface{}
 					collName   interface{}
 					ordinalPos uint64
 					colDefault interface{}
+					charMaxLen interface{}
+					colType    string
 				)
 				if c.Nullable {
 					nullable = "YES"
 				} else {
 					nullable = "NO"
 				}
+				dataType = strings.ToLower(c.Type.String())
 				if sql.IsText(c.Type) {
 					charName = sql.Collation_Default.CharacterSet().String()
 					collName = sql.Collation_Default.String()
+					if st, ok := c.Type.(sql.StringType); ok {
+						charMaxLen = st.MaxCharacterLength()
+					}
+					dataType = strings.TrimSuffix(dataType, fmt.Sprintf("(%v)", charMaxLen))
 				}
 				ordinalPos = uint64(i + 1)
 
 				fullColumnName := db.Name() + "." + t.Name() + "." + c.Name
 				colDefault = trimColumnDefaultOutput(columnNameToDefault[fullColumnName])
+				colType = strings.ToLower(c.Type.String())
+				if c.Type == sql.Boolean {
+					colType = colType + "(1)"
+				}
 				columnKey := ""
+				// PK is checked here first because there are PKs from table implementations that don't implement sql.IndexedTable
 				if c.PrimaryKey {
 					columnKey = "PRI"
+				} else if val, ok := columnKeyMap[c.Name]; ok {
+					columnKey = val
+					// A UNIQUE index may be displayed as PRI if it cannot contain NULL values and there is no PRIMARY KEY in the table
+					if !c.Nullable && !hasPK && columnKey == "UNI" {
```

Comment:
It looks like MySQL will only assign `PRI` to the first column that is not nullable and unique in a keyless table; after that, the other columns get `UNI` for their column key attribute.